pkg/resourcetracker/app.go

```diff
@@ -0,0 +1,129 @@
+/*
+Copyright 2021 The KubeVela Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package resourcetracker
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/crossplane/crossplane-runtime/pkg/meta"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/oam-dev/kubevela/apis/core.oam.dev/v1beta1"
+	"github.com/oam-dev/kubevela/pkg/oam"
+)
+
+const (
+	// Finalizer for resourcetracker to clean up recorded resources
+	Finalizer = "resourcetracker.core.oam.dev/finalizer"
+)
+
+func getRootResourceTrackerName(app *v1beta1.Application) string {
+	return fmt.Sprintf("%s-%s", app.Name, app.Namespace)
+}
+
+func getCurrentResourceTrackerName(app *v1beta1.Application) string {
+	return fmt.Sprintf("%s-v%d-%s", app.Name, app.GetGeneration(), app.Namespace)
+}
```

Comment:
Should consider a situation that user specifies the publish version

Reply:
Ok.